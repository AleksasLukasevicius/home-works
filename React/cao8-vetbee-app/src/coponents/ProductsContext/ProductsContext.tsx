import { createContext } from "vm";

export const ProductsContext = createContext();
