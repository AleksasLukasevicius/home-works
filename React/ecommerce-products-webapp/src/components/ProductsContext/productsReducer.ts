import { TProductsAction, TProductsState } from "./types";

export const productsReducer = (
  state: TProductsState,
  action: TProductsAction
) => {
  switch (action.type) {
    case "addProduct":
      console.log("add product");
      break;

    case "deleteProduct":
      console.log("delete product");
      break;

    case "setProducts":
      console.log("set products");
      break;

    default:
      console.log("no case matched");
  }
  return state;
};
