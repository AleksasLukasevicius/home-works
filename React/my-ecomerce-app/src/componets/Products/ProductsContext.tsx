import { createContext } from "react";
import { TProduct } from "../Types/TProduct";

export const ProductsContext = createContext<TProduct[]>([]);
