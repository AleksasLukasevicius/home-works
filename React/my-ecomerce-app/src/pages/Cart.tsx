import { useContext } from "react";
import { OrangeButton, WhiteButton } from "../componets/Button/Button.styled";
import { CartProductsContext } from "../componets/CartProductsContext/CartProductsContext";

export const Cart = () => {
  const { cartProducts, setCartProducts } = useContext(CartProductsContext);

  const handleAdd = (productIndex: number) => {
    const modifiedProducts = [...cartProducts];
    const product = modifiedProducts[productIndex];

    product.amount += 1;

    setCartProducts(modifiedProducts);
  };

  const handleRemove = (productIndex: number) => {
    const modifiedProducts = [...cartProducts];
    const product = modifiedProducts[productIndex];

    product.amount -= 1;

    if (!product.amount) {
      cartProducts.splice(productIndex, 1);

      return setCartProducts([...cartProducts]);
    }
    setCartProducts(modifiedProducts);
  };

  return (
    <main>
      <div className="title-wrapper">
        <h1>Cart Product List</h1>
      </div>

      <div className="products-container">
        {cartProducts.map((product, productIndex) => (
          <div className="product-card" key={`product.id- ${productIndex}`}>
            <p>Name: {product.name}</p>
            <p>Product amount: {product.amount}</p>
            {product.price ? (
              <>
                <p>Price: {product.price} $</p>
                <p>Sum: {(product.price * product.amount).toFixed(2)} $</p>{" "}
              </>
            ) : null}
            <div className="button-wrapper">
              <OrangeButton onClick={() => handleAdd(productIndex)}>
                Add product
              </OrangeButton>
              <WhiteButton onClick={() => handleRemove(productIndex)}>
                Remove product
              </WhiteButton>
            </div>
          </div>
        ))}
      </div>
    </main>
  );
};
