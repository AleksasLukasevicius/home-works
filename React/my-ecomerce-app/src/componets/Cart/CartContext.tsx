import { createContext } from "react";
import { TCartProductsContext } from "../ProductsContext/types";

export const CartContext = createContext<TCartProductsContext>({
  cartproducts: [],
  setCartProducts: () => {},
});
