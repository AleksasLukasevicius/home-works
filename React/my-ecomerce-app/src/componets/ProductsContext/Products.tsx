import { useContext } from "react";
import type { TProduct } from "../types/TProduct";
import { ProductCard } from "./ProductCard";
import { ProductsContext } from "./ProductsContext";

export const Products = () => {
  const { products, setProducts } = useContext(ProductsContext);
  // const { cartProducts, setCartProducts } = useContext(CartProductsContext);

  const handleAddToCart = (product: TProduct, productIndex: number) => {
    const modifyProducts = [...products];

    modifyProducts[productIndex] = {
      ...product,
      amount: product.amount + 1,
    };
    // modifyProducts[productIndex] = {
    //   ...modifyProducts[productIndex],
    //   amount: modifyProducts[productIndex].amount + 1,
    // };

    const newPrducts = products.map((curProduct) => {
      const isSelectProduct = curProduct.id === product.id;

      return {
        ...curProduct,
        amount: isSelectProduct ? curProduct.amount + 1 : curProduct.amount,
      };
    });
    setProducts(modifyProducts);
  };

  return (
    <main>
      <div className="title-wrapper">
        <h1>Product List</h1>
      </div>

      <div className="products-container">
        {products.map((product, productIndex) => (
          <div className="product-card" key={`product.id- ${productIndex}`}>
            <p>Name: {product.name}</p>
            <p>Price: {product.price} $</p>
            <button
              type="submit"
              onClick={() => handleAddToCart(product, productIndex)}
            >
              Add to Cart
            </button>
          </div>
        ))}
      </div>
    </main>
  );
};
