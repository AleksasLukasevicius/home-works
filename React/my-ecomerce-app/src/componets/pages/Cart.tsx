import { useContext } from "react";
import { OrangeButton, WhiteButton } from "../Button/Button.styled";
import { CartProductsContext } from "../CartProductsContext/CartProductsContext";
import { ProductsContext } from "../ProductsContext/ProductsContext";

export const Cart = () => {
  const { cartProducts, setCartProducts } = useContext(CartProductsContext);
  const { products, setProducts } = useContext(ProductsContext);
  return (
    <main>
      <div className="title-wrapper">
        <h1>Cart Product List</h1>
      </div>

      <div className="products-container">
        {cartProducts.map((product, productIndex) => (
          <div className="product-card" key={`product.id- ${productIndex}`}>
            <p>Name: {product.name}</p>
            <p>Product amount: {product.amount}</p>

            <p>Price: {product.price} $</p>
            <div className="button-wrapper">
              <OrangeButton
              // onClick={() => handleAddToCart(product, productIndex)}
              >
                Add product
              </OrangeButton>
              <WhiteButton
              // onClick={() => handleAddToCart(product, productIndex)}
              >
                Remove product
              </WhiteButton>
            </div>
          </div>
        ))}
      </div>
    </main>
  );
};
