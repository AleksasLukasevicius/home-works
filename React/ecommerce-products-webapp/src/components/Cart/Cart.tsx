import { useContext } from "react";
import { ProductsContext } from "../ProductsContext";
import { Box, Grid, Typography } from "@mui/material";
import { ProductActionButon } from "../Products/ProductActionButton";

export const Cart = () => {
  const { cartProducts } = useContext(ProductsContext);
  const CURRENCY = "EUR";
  const currencyFormatter = new Intl.NumberFormat(navigator.language, {
    style: "currency",
    currency: CURRENCY,
  });

  const initialValue = 0;
  const totalPrice = cartProducts.reduce(
    (curPrice, cartProduct) =>
      curPrice + (cartProduct.price || 0) * cartProduct.amount,
    initialValue
  );

  // todo DataGrid

  return (
    <main>
      <Box borderBottom="2px solid black" mb={2}>
        <Typography variant="h1" component="h1">
          Cart products
        </Typography>
      </Box>

      <Box component="ol">
        {cartProducts.map((product) => {
          return (
            <Grid
              component="li"
              container
              height={50}
              alignItems="center"
              justifyContent="space-between"
              mt={2}
              borderBottom="1px solid black"
            >
              <Grid item xs={7}>
                <Typography variant="body1" pl={1}>
                  {product.title}
                </Typography>
              </Grid>

              <Grid
                item
                xs={3}
                display="flex"
                gap={2}
                justifyContent="space-around"
              >
                <ProductActionButon
                  color="success"
                  title="+"
                  type="addProduct"
                  productId={product.id}
                />

                <Typography>{product.amount}</Typography>

                <ProductActionButon
                  color="error"
                  title="-"
                  type="deleteProduct"
                  productId={product.id}
                />
              </Grid>

              <Grid item xs={2}>
                {product.price ? (
                  <Typography variant="body1" textAlign="right" pr={1}>
                    {/* {product.price?.toFixed(2)} {CURRENCY} */}
                    {currencyFormatter.format(product.price)}
                  </Typography>
                ) : null}
              </Grid>
            </Grid>
          );
        })}

        <Box display="flex" justifyContent="space-between" mt="20px">
          <Typography pl={1}>Total Price:&nbsp;</Typography>
          <Typography pr={1}>{currencyFormatter.format(totalPrice)}</Typography>
        </Box>
      </Box>
    </main>
  );
};
