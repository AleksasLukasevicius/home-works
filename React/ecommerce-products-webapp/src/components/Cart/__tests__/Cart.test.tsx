import { render, screen } from "@testing-library/react";
import { Cart } from "../Cart";
import renderer from "react-test-renderer";
import { ProductsContext } from "../../ProductsContext";

describe("Cart", () => {
  it("should display cart information", () => {
    render(
      <ProductsContext.Provider
        value={{ cartProducts: [], fetchedProducts: [], dispatch: () => {} }}
      >
        <Cart />
      </ProductsContext.Provider>
    );

    expect(screen.getByLabelText("cart")).toBeVisible();
    // expect(screen.getByText("Total: 0 â‚¬")).toBeVisible();
    // expect(screen.getByText("Checkout")).toBeVisible();
  });

  it.only("matches cart snapshot", () => {
    const tree = renderer.create(<Cart />);

    expect(tree).toMatchSnapshot();
  });
});
