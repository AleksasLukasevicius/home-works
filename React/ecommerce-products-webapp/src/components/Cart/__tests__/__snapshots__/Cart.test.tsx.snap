// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart matches cart snapshot 1`] = `
<main>
  <div
    className="MuiBox-root css-uu1ipd"
  >
    <h1
      aria-label="cart-title"
      className="MuiTypography-root MuiTypography-h1 css-o2w69a-MuiTypography-root"
    >
      Cart products
    </h1>
  </div>
  <ol
    aria-label="cart-products"
    className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
  >
    <div
      className="MuiBox-root css-1rldm5g"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-13zw5gk-MuiTypography-root"
      >
        Total Price: 
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-1t69itb-MuiTypography-root"
      >
        €0.00
      </p>
    </div>
  </ol>
</main>
`;
